python/glow/conversions.py

```diff
@@ -0,0 +1,104 @@
+    java_arr = sc._gateway.new_array(sc._jvm.double, np_arr.shape[0])
+    for idx, ele in enumerate(np_arr):
+        java_arr[idx] = ele.item()
+
+    assert check_return_type(java_arr)
+    return java_arr
+
+
+class ArrayConverter(object):
+    """
+    Replaces any 1-dimensional numpy array of doubles with a literal Java array.
+
+    Added in version 0.5.0.
+
+    Examples:
+        >>> import numpy as np
+        >>> from pyspark.sql.functions import lit
+        >>> from pyspark.sql.types import StringType
+        >>> str_list = ['a', 'b']
+        >>> df = spark.createDataFrame(str_list, StringType())
+        >>> ndarray = np.array([1.0, 2.1, 3.2])
+        >>> df.withColumn("array", lit(ndarray)).collect()
+        [Row(value='a', array=[1.0, 2.1, 3.2]), Row(value='b', array=[1.0, 2.1, 3.2])]
+    """
+
+    def can_convert(self, object):
+        return _is_numpy_double_array(object, dimensions = 1)
+
+    def convert(self, object):
+        jvm = SparkContext._active_spark_context._jvm
+        java_arr = _convert_numpy_to_java_array(object)
+        arr_dt = jvm.org.apache.spark.sql.types.ArrayType.apply(jvm.org.apache.spark.sql.types.DataTypes.DoubleType)
+        return java_arr, arr_dt
+
+
+class MatrixConverter(object):
+    """
+    Replaces any 2-dimensional numpy array of doubles with a literal DenseMatrix.
+
+    Added in version 0.5.0.
+
+    Examples:
+        >>> import numpy as np
+        >>> from pyspark.sql.functions import lit
+        >>> from pyspark.sql.types import StringType
+        >>> str_list = ['a', 'b']
+        >>> df = spark.createDataFrame(str_list, StringType())
+        >>> ndarray = np.array([[1.0, 2.1, 3.2], [4.3, 5.4, 6.5]])
+        >>> df.withColumn("matrix", lit(ndarray)).collect()
+        [Row(value='a', matrix=DenseMatrix(2, 3, [1.0, 2.1, 3.2, 4.3, 5.4, 6.5], False)), Row(value='b', matrix=DenseMatrix(2, 3, [1.0, 2.1, 3.2, 4.3, 5.4, 6.5], False))]
+    """
+
+    def can_convert(self, object):
+        return _is_numpy_double_array(object, dimensions = 2)
+
+    def convert(self, object):
+        jvm = SparkContext._active_spark_context._jvm
+        flat_arr = object.ravel()
+        java_arr = _convert_numpy_to_java_array(flat_arr)
+        dense_matrix = jvm.org.apache.spark.ml.linalg.DenseMatrix(object.shape[0], object.shape[1], java_arr)
+        matrix_udt = jvm.org.apache.spark.ml.linalg.MatrixUDT()
+        return dense_matrix, matrix_udt
+
+# Perform conversions inside literal function calls
+glow_java_converter_list = [ArrayConverter(), MatrixConverter()]
+def lit_decorator(lit_fn):
+    def decorate_lit(col):
+        for java_converter in glow_java_converter_list:
+            if java_converter.can_convert(col):
+                jvm = SparkContext._active_spark_context._jvm
+                java_obj, data_type = java_converter.convert(col)
+                catalyst_converter = jvm.org.apache.spark.sql.catalyst.CatalystTypeConverters.createToCatalystConverter(
```

Comment:
It might work to call `Literal.create` directly on the Java object without creating a type converter manually.

Reply:
We can't use `create[T : TypeTag](v: T)` due to type tags being filled in by the compiler. However, I believe we also can't natively use `Literal.create(v, dataType)`, which depends on the limited types offered in `convertToCatalyst ` versus those in `createToCatalystConverter`. I changed this to directly call the constructor for `Literal` to match the type-tagged implementation.